### DISCUSSION TITLE
Aggregate server response

### DISCUSSION THREAD

**Discussion by pokusak** [netty/netty#11828] [Q&A] [answered]
Using Spring cloud gateway 2020.0.3 (uses Netty 4.1.67 internally) I would like to log HTTP headers and body (possibly chunked) to single line. To be able to log it I first need it to log at least in one log message. Request is logged in single log message (multiline) on its own, nothing special is needed. Server response on the other hand is logged in multiple log events headers first, some other lines and body in second log event.

Netty reactor team directed me to HttpObjectAggregator which seems to be exactly what I need. But it is called only on requests not responses. Since I don't know how to link Netty reactor Gitter thread allow me to copy&paste my question from there:

According to the documentation handlers in pipeline are called on inbound/outbound based on the fact if handler implements interface: ChannelInboundHandler or ChannelOutboundHandler. If so, adding HttpObjectAggregator won't work with response (despite the fact that its coode looks like it would do something also with responses) because its hierarchy is: HttpObjectAggregator -> MessageAggregator -> MessageToMessageDecoder -> ChannelInboundHandlerAdapter -> ChannelHandlerAdapter, ChannelInboundHandler -> ChannelHandler There is no outbound interface there.

What has to be done to use HttpObjectAggregator also for server response?

Thanks.

**Comment by normanmaurer** [comment, marked as answer]
Actually it works for responses and requests.

**Reply by pokusak** [reply]
OK, I am not saying it is not. It is just not working for me which also corresponds with my understanding of Netty pipeline working. 
If you could elaborate a little on how is it ensured that is is called also for responses? Obviously it is not via implementation of ChannelOutboundHandler. 
beginAggregate method is called from MessageAggregator for request. From where should it be called for reponses?
Thanks.

**Reply by normanmaurer** [reply]
are you are talking about stuff that you write ? I thought you talked about what you read which might be requests or responses based on if its a server or client. If its about what you write then it will not help and you need to write something yourself.

**Reply by pokusak** [reply]
Yes, I need to aggregate server response == something I write out. 
Does it mean that HttpObjectAggregator works only for read operations (read request by server, read response by client) but not for write operations (write request by client, write response by server)?

**Reply by normanmaurer** [reply]
Yes exactly that
